drivers/wasapi/audio_driver_wasapi.cpp

```diff
@@ -732,241 +814,304 @@ void AudioDriverWASAPI::write_sample(WORD format_tag, int bits_per_sample, BYTE
-							}
-						} else {
-							for (unsigned int i = 0; i < write_frames; i++) {
-								for (unsigned int j = 0; j < MIN(ad->channels, ad->audio_output.channels); j++) {
-									ad->write_sample(ad->audio_output.format_tag, ad->audio_output.bits_per_sample, buffer, i * ad->audio_output.channels + j, ad->samples_in.write[write_ofs++]);
-								}
-								if (ad->audio_output.channels > ad->channels) {
-									for (unsigned int j = ad->channels; j < ad->audio_output.channels; j++) {
-										ad->write_sample(ad->audio_output.format_tag, ad->audio_output.bits_per_sample, buffer, i * ad->audio_output.channels + j, 0);
-									}
-								}
-							}
-						}
-
-						hr = ad->audio_output.render_client->ReleaseBuffer(write_frames, 0);
-						if (hr != S_OK) {
-							ERR_PRINT("WASAPI: Release buffer error");
-						}
-
-						avail_frames -= write_frames;
-						written_frames += write_frames;
-					} else if (hr == AUDCLNT_E_DEVICE_INVALIDATED) {
-						// output_device is not valid anymore, reopen it
-
-						Error err = ad->finish_output_device();
-						if (err != OK) {
-							ERR_PRINT("WASAPI: finish_output_device error");
-						} else {
-							// We reopened the output device and samples_in may have resized, so invalidate the current avail_frames
-							avail_frames = 0;
-						}
+					avail_frames -= write_frames;
+					written_frames += write_frames;
+				} else if (hr == AUDCLNT_E_DEVICE_INVALIDATED) {
+					// output_device is not valid anymore, reopen it
+					Error err = finish_output_device();
+					if (err != OK) {
+						ERR_PRINT("WASAPI: finish_output_device error");
 					} else {
-						ERR_PRINT("WASAPI: Get buffer error");
-						ad->exit_thread.set();
+						// We reopened the output device and samples_in may have resized, so invalidate the current avail_frames
+						avail_frames = 0;
 					}
+				} else {
+					ERR_PRINT("WASAPI: Get buffer error");
+					exit_thread.set();
 				}
-			} else if (hr == AUDCLNT_E_DEVICE_INVALIDATED) {
-				invalidated = true;
-			} else {
-				ERR_PRINT("WASAPI: GetCurrentPadding error");
 			}
+		} else if (hr == AUDCLNT_E_DEVICE_INVALIDATED) {
+			invalidated = true;
+		} else {
+			ERR_PRINT("WASAPI: GetCurrentPadding error");
+		}
 
-			if (invalidated) {
-				// output_device is not valid anymore
-				WARN_PRINT("WASAPI: Current output_device invalidated, closing output_device");
+		if (invalidated) {
+			// output_device is not valid anymore
+			WARN_PRINT("WASAPI: Current output_device invalidated, closing output_device");
 
-				Error err = ad->finish_output_device();
-				if (err != OK) {
-					ERR_PRINT("WASAPI: finish_output_device error");
-				}
+			Error err = finish_output_device();
+			if (err != OK) {
+				ERR_PRINT("WASAPI: finish_output_device error");
 			}
+			// Reset avail_frames to force regeneration of audio data with new buffer size
+			avail_frames = 0;
```

Comment:
This line is added. Can you explain a bit more why it is necessary to reset avail_frames in this case, given that the audio device is closed? The application may still want to read whatever frames would have been available.